File: vox/swift/rapace-runtime/Sources/RapaceRuntime/COBS.swift
import Foundation

// MARK: - COBS encoding/decoding

/// Encode bytes using COBS (Consistent Overhead Byte Stuffing).
/// r[impl message.cobs]
public func cobsEncode(_ input: [UInt8]) -> [UInt8] {
    var out: [UInt8] = []
    out.reserveCapacity(input.count + 2)

    var codeIndex = 0
    var code: UInt8 = 1
    out.append(0) // placeholder

    for b in input {
        if b == 0 {
            out[codeIndex] = code
            codeIndex = out.count
            out.append(0) // placeholder
            code = 1
        } else {
            out.append(b)
            code &+= 1
            if code == 0xFF {
                out[codeIndex] = code
                codeIndex = out.count
                out.append(0)
                code = 1
            }
        }
    }

    out[codeIndex] = code
    return out
}

/// Decode COBS-encoded bytes.
/// r[impl message.cobs]
public func cobsDecode(_ input: [UInt8]) throws -> [UInt8] {
    var out: [UInt8] = []
    out.reserveCapacity(input.count)

    var i = 0
    while i < input.count {
        let code = input[i]
        i += 1
        guard code != 0 else {
            throw RapaceError.decodeError("cobs: zero code byte")
        }
        let n = Int(code) - 1
        guard i + n <= input.count else {
            throw RapaceError.decodeError("cobs: data overrun")
        }
        if n > 0 {
            out.append(contentsOf: input[i..<(i + n)])
            i += n
        }
        if code != 0xFF && i < input.count {
            out.append(0)
        }
    }

    return out
}

File: vox/typescript/packages/roam-core/src/binary/cobs.ts
// COBS encoding/decoding (minimal, no external deps).
export function cobsEncode(input: Uint8Array): Uint8Array {
  const out: number[] = [];
  let codeIndex = 0;
  let code = 1;
  out.push(0); // code placeholder

  for (let i = 0; i < input.length; i++) {
    const b = input[i]!;
    if (b === 0) {
      out[codeIndex] = code;
      codeIndex = out.length;
      out.push(0);
      code = 1;
    } else {
      out.push(b);
      code++;
      if (code === 0xff) {
        out[codeIndex] = code;
        codeIndex = out.length;
        out.push(0);
        code = 1;
      }
    }
  }

  out[codeIndex] = code;
  return Uint8Array.from(out);
}

export function cobsDecode(input: Uint8Array): Uint8Array {
  const out: number[] = [];
  let i = 0;
  while (i < input.length) {
    const code = input[i++]!;
    if (code === 0) throw new Error("cobs: zero code");
    const n = code - 1;
    if (i + n > input.length) throw new Error("cobs: overrun");
    for (let j = 0; j < n; j++) out.push(input[i++]!);
    if (code !== 0xff && i < input.length) out.push(0);
  }
  return Uint8Array.from(out);
}

